File: sample-videochat-webrtc/sample-videochat-webrtc/Views/QBToolBar/QBButtonsFactory.m
//
//  QBButtonsFactory.m
//  sample-videochat-webrtc
//
//  Created by Andrey Ivanov on 23/10/15.
//  Copyright © 2015 QuickBlox Team. All rights reserved.
//

#import "QBButtonsFactory.h"
#import "QBButton.h"

const CGRect kDefRect = {0, 0, 44, 44};
const CGRect kDefDeclineRect = {0, 0, 96, 44};
const CGRect kDefCircleDeclineRect = {0, 0, 44, 44};

#define kDefBackgroundColor [UIColor colorWithRed:0.8118 green:0.8118 blue:0.8118 alpha:1.0]
#define kDefSelectedColor [UIColor colorWithRed:0.3843 green:0.3843 blue:0.3843 alpha:1.0]
#define kDefDeclineColor [UIColor colorWithRed:0.8118 green:0.0 blue:0.0784 alpha:1.0]
#define kDefAnswerColor [UIColor colorWithRed:0.1434 green:0.7587 blue:0.1851 alpha:1.0]

@implementation QBButtonsFactory

#pragma mark - Private

+ (QBButton *)buttonWithFrame:(CGRect)frame
              backgroundColor:(UIColor *)backgroundColor
                selectedColor:(UIColor *)selectedColor  {
    
    QBButton *button = [[QBButton alloc] initWithFrame:frame];
    button.backgroundColor = backgroundColor;
    button.selectedColor = selectedColor;
    
    return button;
}

+ (UIImageView *)iconViewWithNormalImage:(NSString *)normalImage
                           selectedImage:(NSString *)selectedImage {
    
    UIImage *icon = [UIImage imageNamed:normalImage];
    UIImage *selectedIcon = [UIImage imageNamed:selectedImage];
    
    UIImageView *iconView = [[UIImageView alloc] initWithImage:icon
                                              highlightedImage:selectedIcon];
    
    iconView.contentMode = UIViewContentModeScaleAspectFit;

    return iconView;
}

#pragma mark - Public

+ (QBButton *)videoEnable {
    
    QBButton *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    button.pushed = YES;
    
    button.iconView = [self iconViewWithNormalImage:@"camera_on_ic"
                                      selectedImage:@"camera_off_ic"];
    return button;
}

+ (QBButton *)auidoEnable {
    
    QBButton *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.pushed = YES;
    
    button.iconView = [self iconViewWithNormalImage:@"mute_on_ic"
                                      selectedImage:@"mute_off_ic"];
    return button;
}

+ (QBButton *)dynamicEnable {
    
    QBButton *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.pushed = YES;
    
    button.iconView = [self iconViewWithNormalImage:@"dynamic_on_ic"
                                      selectedImage:@"dynamic_off_ic"];
    return button;
}

+ (QBButton *)screenShare {
    
    QBButton *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"screensharing_ic"
                                      selectedImage:@"screensharing_ic"];
    return button;
}

+ (QBButton *)answer {
    
    QBButton *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefAnswerColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"answer"
                                      selectedImage:@"answer"];
    return button;
}

+ (QBButton *)decline {
    
    QBButton *button = [self buttonWithFrame:kDefDeclineRect
                             backgroundColor:kDefDeclineColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"decline-ic"
                                      selectedImage:@"decline-ic"];
    return button;
}

+ (QBButton *)circleDecline {
    
    QBButton *button = [self buttonWithFrame:kDefCircleDeclineRect
                             backgroundColor:kDefDeclineColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"decline-ic"
                                      selectedImage:@"decline-ic"];
    return button;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/QBToolBar/QBButtonsFactory.swift
//
//  QBButtonsFactory.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

struct QBButtonsFactoryConstants {
    static let defRect: CGRect = CGRect(x: 0, y: 0, width: 44, height: 44)
    static let defDeclineRect: CGRect = CGRect(x: 0, y: 0, width: 96, height: 44)
    static let defCircleDeclineRect: CGRect = CGRect(x: 0, y: 0, width: 44, height: 44)
    
    static let defBackgroundColor = UIColor(red: 0.8118, green: 0.8118, blue: 0.8118, alpha: 1.0)
    static let defSelectedColor = UIColor(red: 0.3843, green: 0.3843, blue: 0.3843, alpha: 1.0)
    static let defDeclineColor = UIColor(red: 0.8118, green: 0.0, blue: 0.0784, alpha: 1.0)
    static let defAnswerColor = UIColor(red: 0.1434, green: 0.7587, blue: 0.1851, alpha: 1.0)
}

class QBButtonsFactory {
    // MARK: - Private
    class func button(withFrame frame: CGRect, backgroundColor: UIColor?, selectedColor: UIColor?) -> QBButton {
        
        let button = QBButton(frame: frame)
        button.backgroundColor = backgroundColor
        button.selectedColor = selectedColor
        
        return button
    }
    
    class func iconView(withNormalImage normalImage: String?, selectedImage: String?) -> UIImageView? {
        
        let icon = UIImage(named: normalImage ?? "")
        let selectedIcon = UIImage(named: selectedImage ?? "")
        
        let iconView = UIImageView(image: icon, highlightedImage: selectedIcon)
        
        iconView.contentMode = .scaleAspectFit
        
        return iconView
    }
    
    // MARK: - Public
    class func videoEnable() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defRect, backgroundColor:QBButtonsFactoryConstants.defBackgroundColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "camera_on_ic", selectedImage: "camera_off_ic")
        return button
    }
    
    class func auidoEnable() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defRect, backgroundColor: QBButtonsFactoryConstants.defBackgroundColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "mute_on_ic", selectedImage: "mute_off_ic")
        return button
    }
    
    class func dynamicEnable() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defRect, backgroundColor: QBButtonsFactoryConstants.defBackgroundColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "dynamic_on_ic", selectedImage: "dynamic_off_ic")
        return button
    }
    
    class func screenShare() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defRect, backgroundColor: QBButtonsFactoryConstants.defBackgroundColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.iconView = self.iconView(withNormalImage: "screensharing_ic", selectedImage: "screensharing_ic")
        return button
    }
    
    class func answer() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defRect, backgroundColor: QBButtonsFactoryConstants.defAnswerColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.iconView = self.iconView(withNormalImage: "answer", selectedImage: "answer")
        return button
    }
    
    class func decline() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defDeclineRect, backgroundColor: QBButtonsFactoryConstants.defDeclineColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.iconView = self.iconView(withNormalImage: "decline-ic", selectedImage: "decline-ic")
        return button
    }
    
    class func circleDecline() -> QBButton {
        
        let button: QBButton = self.button(withFrame: QBButtonsFactoryConstants.defCircleDeclineRect, backgroundColor: QBButtonsFactoryConstants.defDeclineColor, selectedColor: QBButtonsFactoryConstants.defSelectedColor)
        button.iconView = self.iconView(withNormalImage: "decline-ic", selectedImage: "decline-ic")
        return button
    }
}
